File: Next Greater Node in Linked list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */


class Solution {
public:

    
    vector<int> nextLargerNodes(ListNode* head) {

        ListNode *temp = head;
        ListNode *temp2 = head->next;
        vector<int> values;
        while(temp != NULL)
        {
            temp2 = temp->next;
            while(temp2 != NULL)
            {
                if(temp2->val > temp->val)
                {
                    values.push_back(temp2->val);
                    break;
                }
                if(temp2->next == NULL)
                    values.push_back(0);
                temp2 = temp2->next;
            }
            if(temp->next == NULL)
                values.push_back(0);
            temp = temp->next;
        }
        return values;
    }
};

File: NextGreaterNodeInLinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def nextLargerNodes(self, head: Optional[ListNode]) -> List[int]:
        curr = head
        prev = None
        stack = []
        sol = []
        # reverse the linked list
        while curr != None:
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next

        head = prev
        
        while head != None:

            while len(stack) != 0 and stack[-1] <= head.val:
                stack.pop()

            if len(stack) == 0:
                sol.append(0)
            else:
                sol.append(stack[-1])

            stack.append(head.val)
            head = head.next

        sol.reverse()
        
        return sol
